Migrating to modest queue

mod35 opened the pull request:
Proposing moving from bullmq to a much simpler queue implementation: https://gitlab.com/ersutton/modest-queue

had some issues with bullmq when processing messages, when you set a job to be completed, it would automatically pop the next message to be processed off the queue - which would have made the implementation in this library messy.

This MR
* migrates to modest-queue
* publishes messages to any queue that has stated that it wants to know about a certain message

Tests pass:
![image](https://user-images.githubusercontent.com/18492498/130465318-7d91502b-54fa-46f2-a00c-8a92e808f7f8.png)

adenhertog commented on packages/bus-redis/src/redis-transport-configuration.ts:
```diff
@@ -21,4 +21,20 @@ export interface RedisTransportConfiguration {
+
+  /**
+   * The time taken before a message has been deemed to have failed or stalled. Once this time has been exceeded
+   * the message will be re-added to the queue.
```
can this clarify if it's added to its original position (made visible), or added as a new message on the end of the queue?

mod35 replied:
Its re-added to the end of the queue

adenhertog commented on packages/bus-redis/src/redis-transport-configuration.ts:
```diff
@@ -21,4 +21,20 @@ export interface RedisTransportConfiguration {
+  visibilityTimeout?: number
+
+  /**
+   * The scheduler is a worker that checks messages if any messages have exceeded thy
```
I didn't quite understand what this does from a consumer perspective. If this is false, does it mean messages aren't retried?

mod35 replied:
Theres a while loop in the queue that checks for messages that have been popped off the queue to see if they have succeeded or failed within the visibility timeout. This `scheduler` might get a bit heavy if EVERY worker thats popping messages off the queue is also checking to see if they have exceeded their visibility timeout - I think this is somewhat easier to understand if you look at the queue itself.
![image](https://user-images.githubusercontent.com/18492498/132110013-6ffceb67-291f-4c80-b2f3-5173b221cfec.png)

mod35 replied:
the file to look at this is here: https://gitlab.com/ersutton/modest-queue/-/blob/main/src/modest-queue.ts

adenhertog commented on packages/bus-redis/src/redis-transport.ts:
```diff
@@ -19,62 +19,78 @@ interface Payload {
+      queueName: this.configuration.queueName,
+      connectionString: this.configuration.connectionString,
+      visibilityTimeout: this.configuration.visibilityTimeout ?? 30000,
+      maxAttempts: 3,
```
should this be

```suggestion
      maxAttempts: this.configuration.maxRetries ?? 3,
```

adenhertog commented on packages/bus-redis/src/redis-transport.ts:
```diff
@@ -166,10 +158,13 @@ export class RedisTransport implements Transport<RedisMessage> {
-      attempts: this.maxRetries,
-      removeOnComplete: true
-    })
+    const serializedPayload = JSON.stringify(payload)
```
this should use `MessageSerializer` from `@node-ts/bus-core`

adenhertog commented on packages/bus-redis/src/redis-transport.ts:
```diff
@@ -166,10 +158,13 @@ export class RedisTransport implements Transport<RedisMessage> {
+    await Promise.all(queues.map(async queueName => {
+      const queue = new ModestQueue({queueName, connection: this.connection, withScheduler: false, withDelayedScheduler: false})
+      await queue.initialize()
+      await queue.publish(serializedPayload)
```
if the payload is already serialized, why in modest-queue does it serialize it again?

![image](https://user-images.githubusercontent.com/6031613/131833745-fc4d8c58-64e0-484b-9718-1988346fbdc6.png)

adenhertog commented on packages/bus-redis/src/redis-transport.ts:
```diff
@@ -86,73 +102,49 @@ export class RedisTransport implements Transport<RedisMessage> {
-    this.logger.debug('Received message from Redis', {redisMessage: job.data})
-    const { message, ...attributes}: Payload = job.data
+    this.logger.debug('Received message from Redis', {redisMessage: maybeMessage.message})
+    const { message, ...attributes}: Payload = JSON.parse(maybeMessage.message)
```
should use `MessageSerializer` from `@node-ts/bus-core`

mod35 replied:
I do use the MessageSerializer for the DomainMessage - but I need to also serialize the other bits - need a bit more guidance for this

adenhertog commented on packages/bus-redis/src/redis-transport.ts:
```diff
@@ -166,10 +158,13 @@ export class RedisTransport implements Transport<RedisMessage> {
-    })
+    const serializedPayload = JSON.stringify(payload)
+    const queues = await this.getQueuesSubscribedToMessage(message)
+    await Promise.all(queues.map(async queueName => {
```
all settled might be a better option so that a failure doesn't affect others. 

```suggestion
    await Promise.allSettled(queues.map(async queueName => {
```

might need to have a think what a failure means in this case. would we want to throw an error, retry and therefore double-send to the other queues? 

maybe here we just want to retry the send to the failed subscriber a couple of times, log an error and give up if it's undeliverable.

mod35 replied:
Good call, I've added a recursive retry approach - should be tail call optimised as the retry call is in the catch block.

mod35 replied:
@andrew have added the allSettled approach - this is ready for another look :pray:

adenhertog commented:
clever use of sets to imitate queues. I had another read of redis's native pub/sub but looks like it's something not at all similar to a message bus. shame as I thought it would have had a native topic we could use but guess seems like your implementation is filling the gap

mod35 commented:
@adenhertog thanks for the great feedback - I'd like to show you around the queue lib itself at some point. Explain why I'm using sets and not just lists (its to do with having built in priority queues that I want to use - multiple queues to get around my personal use case won't cut it). I'd appreciate feedback any feedback on that too :)

mod35 commented on packages/bus-redis/src/redis-transport.ts:
```diff
@@ -166,10 +158,13 @@ export class RedisTransport implements Transport<RedisMessage> {
+    const serializedPayload = JSON.stringify(payload)
+    const queues = await this.getQueuesSubscribedToMessage(message)
+    await Promise.all(queues.map(async queueName => {
+      const queue = new ModestQueue({queueName, connection: this.connection, withScheduler: false, withDelayedScheduler: false})
```
exactly. overly verbose explanation - your lib doesn't need delayed support. and we aren't going to be polling any messages here - we're just posting them onto the queue so we don't need to check if they have exceeded their inflight time.

mod35 commented:
@adenhertog I've updated the modest-queue version to 0.4.1 which:
1. respects the placement of messages in the queue when retrying them (no longer places them to the right of the queue - its the equivalent of the message becoming visible again)
2. I've also updated the tsconfig in the root of the project because allSettled was required - this should have been fine the bus-redis projects tsconfig, but tests failed (because I imagine I'd need a separate tsconfig for jest so that it has allSettled. Interested with your thoughts on this.

mod35 commented on tsconfig.json:
```diff
@@ -5,6 +5,7 @@
     "paths": {
       "@node-ts/*": ["./*/src"]
     },
+    "lib": ["ES2020"],
```
I mentioned the reason for this in the comment on the MR - happy to chat on this or if you can suggest a better approach.
